File: Reverse Vowels of a String.java
// https://leetcode.com/problems/reverse-vowels-of-a-string/

class Solution {
    public String reverseVowels(String s) {
        
        // If string is null or only has 1 letter, return it
        if (s == null || s.length() == 1) {
            return s;
        }
        
        // Create 2 pointers
        int leftPtr = 0;
        int rightPtr = s.length() - 1;
        
        // Convert string to a list
        char[] sList = s.toCharArray();
        
        // Create a list of vowels
        List<Character> vowels = new ArrayList<Character>();
        vowels.add('A');
        vowels.add('a');
        vowels.add('E');
        vowels.add('e');
        vowels.add('I');
        vowels.add('i');
        vowels.add('O');
        vowels.add('o');
        vowels.add('U');
        vowels.add('u');
        
        // Iterate list
        while (leftPtr < rightPtr) {
            
            // If both pointers are at a vowel, swap them
            if (vowels.contains(sList[leftPtr]) && vowels.contains(sList[rightPtr])) {
                
                // Swap values
                char oldLetter = sList[leftPtr];
                sList[leftPtr] = sList[rightPtr];
                sList[rightPtr] = oldLetter;
                
                // Move pointers
                leftPtr++;
                rightPtr--;
            }
            
            // If leftPtr is at a vowel, but rightPtr isn't, move rightPtr
            else if (vowels.contains(sList[leftPtr]) && !vowels.contains(sList[rightPtr])) {
                rightPtr--;
            }
            
            // If rightPtr is at a vowel, but leftPtr isn't, move leftPtr
            else if (!vowels.contains(sList[leftPtr]) && vowels.contains(sList[rightPtr])) {
                leftPtr++;
            }
            
            // Move both pointers
            else {
                leftPtr++;
                rightPtr--;
            }
        }
        return new String(sList);  
    }
}

File: Reverse Vowels of a String.py
# https://leetcode.com/problems/reverse-vowels-of-a-string/

class Solution:
    def reverseVowels(self, s: str) -> str:
        
        # If s is none or only has a length of 1, return it
        if s is None or len(s) == 1:
            return s
        
        # Convert all letters to lowercase
        s = list(s)
        
        # Create 2 pointers
        leftPtr = 0
        rightPtr = len(s) - 1
        
        # Iterate string
        vowels = {"a", "A", "e", "E", "i", "I", "o", "O", "u", "U"}
        while leftPtr < rightPtr:
            
            # Check that both pointers are at vowels
            if s[leftPtr] in vowels and s[rightPtr] in vowels:
                
                # Swap letters
                oldLetter = s[leftPtr]
                s[leftPtr] = s[rightPtr]
                s[rightPtr] = oldLetter
                
                # Move pointers
                leftPtr += 1
                rightPtr -= 1
        
            # If rightPtr isn't at a vowel, move it
            elif s[leftPtr] in vowels and s[rightPtr] not in vowels:
                rightPtr -= 1
            
            # If leftPtr isn't at a vowel, move it
            elif s[leftPtr] not in vowels and s[rightPtr] in vowels:
                leftPtr += 1
            
            # If neither pointers are at a vowel, move pointers
            else:
                leftPtr += 1
                rightPtr -= 1
        
        # Convert list back to string
        return "".join(s)
